## 0. Before we begin....

**Reference**
* https://github.com/BrainOmicsCourse/BrainOmics2024/blob/main/1_Day1/Compiled/0_AssembleAdata.html
* https://github.com/BrainOmicsCourse/BrainOmics2024/blob/main/1_Day1/Resources.md
* https://scbrainregulation.su.domains/
* https://www.cell.com/cell/fulltext/S0092-8674(21)00942-9?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867421009429%3Fshowall%3Dtrue

## 1. Python Environment Setup
|name|version|
|---|---|
|macOS||
|Python| 3.9.7|
|numpy|2.0.2|
|pandas|2.2.2|
|matplotlib|3.9.4|
|scanpy|1.10.1|

### 1.1. Download Container
!will be updated

### 1.2. If the container is not working -> please set the libraries above

## 2. Data Download

In [1]:
### Name of Example Files
rna_counts = 'GSE162170_rna_counts.tsv'
rna_meta = 'GSE162170_rna_cell_metadata.txt'

### Data Path
path = '../data/'
Id = 'Id0003'
count_file  = f"{path}/{Id}/{rna_counts}"
meta_file   = f"{path}/{Id}/{rna_meta}"
output_file = f"{path}/{Id}/1_AssembledAdata.h5ad"

In [2]:
### Download Data
!curl -o {rna_counts}.gz https://ftp.ncbi.nlm.nih.gov/geo/series/GSE162nnn/GSE162170/suppl/GSE162170%5Frna%5Fcounts%2Etsv%2Egz
!curl -o {rna_meta}.gz https://ftp.ncbi.nlm.nih.gov/geo/series/GSE162nnn/GSE162170/suppl/GSE162170%5Frna%5Fcell%5Fmetadata%2Etxt%2Egz
# If the download code is unavailable,
# please visit https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE162170
# and download two files below:
#   GSE162170_rna_counts.tsv.gz
#   GSE162170_rna_cell_metadata.txt.gz
# and place them to the proper directory

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  133M  100  133M    0     0  9928k      0  0:00:13  0:00:13 --:--:-- 11.9M  8236k      0  0:00:16  0:00:07  0:00:09 11.4M0:13  0:00:12  0:00:01 12.1M
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 2829k  100 2829k    0     0   875k      0  0:00:03  0:00:03 --:--:--  875k


In [3]:
### Unzip the data
!gzip -d {rna_counts}.gz
!gzip -d {rna_meta}.gz

### Place the data
!mkdir -p {path}/{Id}
!mv {rna_counts} {count_file}
!mv {rna_meta} {meta_file}

## 2. Build Anndata
### 2.1. Preparation

In [4]:
### System
import os
import sys

### Data Analysing
import numpy as np  
import pandas as pd
import matplotlib.pyplot as plt

### Single Cell RNA Sequening (mostly from scverse)
import anndata as ad
import scanpy as sc
from scipy.sparse import csr_matrix, isspmatrix
import scanpy.external as sce

### Custom functions
sys.path.append('../HelperFunctions')
import Helper as fn

In [5]:
print("Scanpy version: ", sc.__version__)
print("Pandas version: ", pd.__version__)

Scanpy version:  1.10.1
Pandas version:  2.2.2


In [6]:
### File Check
print(os.path.exists(count_file))
print(os.path.exists(meta_file))

True
True


### 2.2. Read RNA 

In [7]:
### 'count_file' will be the main of anndata
adata = sc.read_csv(
    count_file,
    delimiter='\t',
    first_column_names=None,
    dtype='float32'
    ) 

In [8]:
### You can choose different strategies when the Anndata is a sparse matrix or not.
### From here, Anndata will be *always* the sparse matrix, and be compressed.
if isspmatrix(adata.X) == False:
    adata.X = csr_matrix(adata.X)
    print('Converted adata.X to', type(adata.X))

Converted adata.X to <class 'scipy.sparse._csr.csr_matrix'>


In [9]:
adata.var

Empty DataFrame
Columns: []
Index: [hft_w20_p3_r1_AAACCCAAGCTGCGAA, hft_w20_p3_r1_AAACCCAAGGTAGTAT, hft_w20_p3_r1_AAACCCACAACTCCAA, hft_w20_p3_r1_AAACCCACATAGTCAC, hft_w20_p3_r1_AAACCCAGTACAGGTG, hft_w20_p3_r1_AAACCCAGTACGGTTT, hft_w20_p3_r1_AAACCCAGTACTCGCG, hft_w20_p3_r1_AAACCCAGTATGTCCA, hft_w20_p3_r1_AAACCCAGTGTATTCG, hft_w20_p3_r1_AAACCCAGTTGCTCAA, hft_w20_p3_r1_AAACCCATCAAGGCTT, hft_w20_p3_r1_AAACCCATCTGCAGCG, hft_w20_p3_r1_AAACCCATCTGTCAGA, hft_w20_p3_r1_AAACGAAAGATGAAGG, hft_w20_p3_r1_AAACGAACAAAGTATG, hft_w20_p3_r1_AAACGAACAGACCATT, hft_w20_p3_r1_AAACGAACAGACCTGC, hft_w20_p3_r1_AAACGAAGTAGCTGTT, hft_w20_p3_r1_AAACGAAGTCACCCTT, hft_w20_p3_r1_AAACGAAGTCGTCATA, hft_w20_p3_r1_AAACGAAGTTGCATCA, hft_w20_p3_r1_AAACGCTGTCATCCCT, hft_w20_p3_r1_AAACGCTTCGTGCTCT, hft_w20_p3_r1_AAAGAACAGAAGCCAC, hft_w20_p3_r1_AAAGAACAGCGACTAG, hft_w20_p3_r1_AAAGAACAGGAGGTTC, hft_w20_p3_r1_AAAGAACAGGTGAGCT, hft_w20_p3_r1_AAAGAACAGTCATTGC, hft_w20_p3_r1_AAAGAACAGTTTCTTC, hft_w20_p3_r1_AAAGAACGTCCAGCGT, hft_w20_p3_r1_AAAGAACGTGCATTTG, hft_w20_p3_r1_AAAGGATAGAGTACCG, hft_w20_p3_r1_AAAGGATAGGTTCAGG, hft_w20_p3_r1_AAAGGATAGTGTAGAT, hft_w20_p3_r1_AAAGGATCAAATGGAT, hft_w20_p3_r1_AAAGGATCATTACTCT, hft_w20_p3_r1_AAAGGGCAGAACAAGG, hft_w20_p3_r1_AAAGGGCCACTTGAAC, hft_w20_p3_r1_AAAGGGCTCACCCTTG, hft_w20_p3_r1_AAAGGGCTCAGCCTCT, hft_w20_p3_r1_AAAGGTAAGGACGCTA, hft_w20_p3_r1_AAAGGTAAGGCATGGT, hft_w20_p3_r1_AAAGGTACATAAGCAA, hft_w20_p3_r1_AAAGGTACATAGTCGT, hft_w20_p3_r1_AAAGGTAGTGTCATCA, hft_w20_p3_r1_AAAGGTATCAGTGGGA, hft_w20_p3_r1_AAAGTCCAGAAGGATG, hft_w20_p3_r1_AAAGTCCAGACTCCGC, hft_w20_p3_r1_AAAGTCCAGCGACTTT, hft_w20_p3_r1_AAAGTCCCAAATTGGA, hft_w20_p3_r1_AAAGTCCCAGTGTACT, hft_w20_p3_r1_AAAGTCCGTAGAATAC, hft_w20_p3_r1_AAAGTCCGTATCGTGT, hft_w20_p3_r1_AAAGTCCGTCAGTCTA, hft_w20_p3_r1_AAAGTGAAGACCTTTG, hft_w20_p3_r1_AAAGTGAAGTTCCGTA, hft_w20_p3_r1_AAAGTGACACTACCCT, hft_w20_p3_r1_AAAGTGACATCATGAC, hft_w20_p3_r1_AAAGTGAGTGACCTGC, hft_w20_p3_r1_AAATGGACACACCTGG, hft_w20_p3_r1_AAATGGACACCTGATA, hft_w20_p3_r1_AAATGGACAGCATTGT, hft_w20_p3_r1_AAATGGAGTAGATGTA, hft_w20_p3_r1_AAATGGAGTGGCGTAA, hft_w20_p3_r1_AAATGGATCTTGCAGA, hft_w20_p3_r1_AACAAAGAGAAGTATC, hft_w20_p3_r1_AACAAAGAGACCTCAT, hft_w20_p3_r1_AACAAAGAGCGAGTCA, hft_w20_p3_r1_AACAAAGCAGGTTACT, hft_w20_p3_r1_AACAAAGCATATGGCT, hft_w20_p3_r1_AACAAAGGTACAAAGT, hft_w20_p3_r1_AACAAAGGTAGCTCGC, hft_w20_p3_r1_AACAAAGGTCAAAGAT, hft_w20_p3_r1_AACAACCAGCAGATAT, hft_w20_p3_r1_AACAACCAGCAGTCTT, hft_w20_p3_r1_AACAACCAGGATACCG, hft_w20_p3_r1_AACAACCCAATGCAAA, hft_w20_p3_r1_AACAACCGTGATTGGG, hft_w20_p3_r1_AACAACCTCATTATCC, hft_w20_p3_r1_AACAACCTCCCATACC, hft_w20_p3_r1_AACAACCTCGTCTACC, hft_w20_p3_r1_AACAAGAAGCCGTTAT, hft_w20_p3_r1_AACAAGACAAACACGG, hft_w20_p3_r1_AACAAGACACATAGCT, hft_w20_p3_r1_AACAAGACACGTCTCT, hft_w20_p3_r1_AACAAGACAGGAGACT, hft_w20_p3_r1_AACAAGAGTCTCAGAT, hft_w20_p3_r1_AACAAGATCGCTAAAC, hft_w20_p3_r1_AACAAGATCGTCCATC, hft_w20_p3_r1_AACACACCACTGCTTC, hft_w20_p3_r1_AACACACCATGACTAC, hft_w20_p3_r1_AACACACGTAGTAAGT, hft_w20_p3_r1_AACACACGTATCTCGA, hft_w20_p3_r1_AACACACTCCTACGAA, hft_w20_p3_r1_AACAGGGAGAGTTGAT, hft_w20_p3_r1_AACAGGGAGATTAGTG, hft_w20_p3_r1_AACAGGGAGCAACAGC, hft_w20_p3_r1_AACAGGGCAGCACAAG, hft_w20_p3_r1_AACAGGGTCGAGAACG, hft_w20_p3_r1_AACAGGGTCTCAATCT, ...]

[57868 rows x 0 columns]

In [10]:
adata.obs

Empty DataFrame
Columns: []
Index: [ENSG00000243485, ENSG00000237613, ENSG00000186092, ENSG00000238009, ENSG00000239945, ENSG00000239906, ENSG00000241599, ENSG00000236601, ENSG00000284733, ENSG00000235146, ENSG00000284662, ENSG00000229905, ENSG00000237491, ENSG00000177757, ENSG00000225880, ENSG00000230368, ENSG00000272438, ENSG00000230699, ENSG00000241180, ENSG00000223764, ENSG00000187634, ENSG00000188976, ENSG00000187961, ENSG00000187583, ENSG00000187642, ENSG00000272512, ENSG00000188290, ENSG00000187608, ENSG00000224969, ENSG00000188157, ENSG00000273443, ENSG00000237330, ENSG00000131591, ENSG00000223823, ENSG00000272141, ENSG00000205231, ENSG00000162571, ENSG00000186891, ENSG00000186827, ENSG00000078808, ENSG00000176022, ENSG00000184163, ENSG00000260179, ENSG00000160087, ENSG00000230415, ENSG00000162572, ENSG00000131584, ENSG00000169972, ENSG00000127054, ENSG00000224051, ENSG00000169962, ENSG00000107404, ENSG00000162576, ENSG00000175756, ENSG00000221978, ENSG00000242485, ENSG00000272455, ENSG00000235098, ENSG00000225905, ENSG00000205116, ENSG00000225285, ENSG00000179403, ENSG00000215915, ENSG00000160072, ENSG00000197785, ENSG00000205090, ENSG00000160075, ENSG00000215014, ENSG00000228594, ENSG00000272106, ENSG00000197530, ENSG00000189409, ENSG00000248333, ENSG00000272004, ENSG00000189339, ENSG00000008128, ENSG00000215790, ENSG00000008130, ENSG00000078369, ENSG00000231050, ENSG00000169885, ENSG00000178821, ENSG00000142609, ENSG00000233542, ENSG00000187730, ENSG00000226969, ENSG00000067606, ENSG00000271806, ENSG00000182873, ENSG00000162585, ENSG00000234396, ENSG00000157933, ENSG00000116151, ENSG00000157916, ENSG00000157911, ENSG00000149527, ENSG00000224387, ENSG00000229393, ENSG00000157881, ENSG00000197921, ...]

[33355 rows x 0 columns]

You can see `adata.var` shows barcodes, while `adata.obs` shows genes.
* We are going to transpose the anndata to replace them
* `obs`(observation) will be barcodes
* `var`(variable) will be genes

In [11]:
### Anndata transpose
adata = adata.transpose()
adata

AnnData object with n_obs × n_vars = 57868 × 33355

In [12]:
adata.obs

Empty DataFrame
Columns: []
Index: [hft_w20_p3_r1_AAACCCAAGCTGCGAA, hft_w20_p3_r1_AAACCCAAGGTAGTAT, hft_w20_p3_r1_AAACCCACAACTCCAA, hft_w20_p3_r1_AAACCCACATAGTCAC, hft_w20_p3_r1_AAACCCAGTACAGGTG, hft_w20_p3_r1_AAACCCAGTACGGTTT, hft_w20_p3_r1_AAACCCAGTACTCGCG, hft_w20_p3_r1_AAACCCAGTATGTCCA, hft_w20_p3_r1_AAACCCAGTGTATTCG, hft_w20_p3_r1_AAACCCAGTTGCTCAA, hft_w20_p3_r1_AAACCCATCAAGGCTT, hft_w20_p3_r1_AAACCCATCTGCAGCG, hft_w20_p3_r1_AAACCCATCTGTCAGA, hft_w20_p3_r1_AAACGAAAGATGAAGG, hft_w20_p3_r1_AAACGAACAAAGTATG, hft_w20_p3_r1_AAACGAACAGACCATT, hft_w20_p3_r1_AAACGAACAGACCTGC, hft_w20_p3_r1_AAACGAAGTAGCTGTT, hft_w20_p3_r1_AAACGAAGTCACCCTT, hft_w20_p3_r1_AAACGAAGTCGTCATA, hft_w20_p3_r1_AAACGAAGTTGCATCA, hft_w20_p3_r1_AAACGCTGTCATCCCT, hft_w20_p3_r1_AAACGCTTCGTGCTCT, hft_w20_p3_r1_AAAGAACAGAAGCCAC, hft_w20_p3_r1_AAAGAACAGCGACTAG, hft_w20_p3_r1_AAAGAACAGGAGGTTC, hft_w20_p3_r1_AAAGAACAGGTGAGCT, hft_w20_p3_r1_AAAGAACAGTCATTGC, hft_w20_p3_r1_AAAGAACAGTTTCTTC, hft_w20_p3_r1_AAAGAACGTCCAGCGT, hft_w20_p3_r1_AAAGAACGTGCATTTG, hft_w20_p3_r1_AAAGGATAGAGTACCG, hft_w20_p3_r1_AAAGGATAGGTTCAGG, hft_w20_p3_r1_AAAGGATAGTGTAGAT, hft_w20_p3_r1_AAAGGATCAAATGGAT, hft_w20_p3_r1_AAAGGATCATTACTCT, hft_w20_p3_r1_AAAGGGCAGAACAAGG, hft_w20_p3_r1_AAAGGGCCACTTGAAC, hft_w20_p3_r1_AAAGGGCTCACCCTTG, hft_w20_p3_r1_AAAGGGCTCAGCCTCT, hft_w20_p3_r1_AAAGGTAAGGACGCTA, hft_w20_p3_r1_AAAGGTAAGGCATGGT, hft_w20_p3_r1_AAAGGTACATAAGCAA, hft_w20_p3_r1_AAAGGTACATAGTCGT, hft_w20_p3_r1_AAAGGTAGTGTCATCA, hft_w20_p3_r1_AAAGGTATCAGTGGGA, hft_w20_p3_r1_AAAGTCCAGAAGGATG, hft_w20_p3_r1_AAAGTCCAGACTCCGC, hft_w20_p3_r1_AAAGTCCAGCGACTTT, hft_w20_p3_r1_AAAGTCCCAAATTGGA, hft_w20_p3_r1_AAAGTCCCAGTGTACT, hft_w20_p3_r1_AAAGTCCGTAGAATAC, hft_w20_p3_r1_AAAGTCCGTATCGTGT, hft_w20_p3_r1_AAAGTCCGTCAGTCTA, hft_w20_p3_r1_AAAGTGAAGACCTTTG, hft_w20_p3_r1_AAAGTGAAGTTCCGTA, hft_w20_p3_r1_AAAGTGACACTACCCT, hft_w20_p3_r1_AAAGTGACATCATGAC, hft_w20_p3_r1_AAAGTGAGTGACCTGC, hft_w20_p3_r1_AAATGGACACACCTGG, hft_w20_p3_r1_AAATGGACACCTGATA, hft_w20_p3_r1_AAATGGACAGCATTGT, hft_w20_p3_r1_AAATGGAGTAGATGTA, hft_w20_p3_r1_AAATGGAGTGGCGTAA, hft_w20_p3_r1_AAATGGATCTTGCAGA, hft_w20_p3_r1_AACAAAGAGAAGTATC, hft_w20_p3_r1_AACAAAGAGACCTCAT, hft_w20_p3_r1_AACAAAGAGCGAGTCA, hft_w20_p3_r1_AACAAAGCAGGTTACT, hft_w20_p3_r1_AACAAAGCATATGGCT, hft_w20_p3_r1_AACAAAGGTACAAAGT, hft_w20_p3_r1_AACAAAGGTAGCTCGC, hft_w20_p3_r1_AACAAAGGTCAAAGAT, hft_w20_p3_r1_AACAACCAGCAGATAT, hft_w20_p3_r1_AACAACCAGCAGTCTT, hft_w20_p3_r1_AACAACCAGGATACCG, hft_w20_p3_r1_AACAACCCAATGCAAA, hft_w20_p3_r1_AACAACCGTGATTGGG, hft_w20_p3_r1_AACAACCTCATTATCC, hft_w20_p3_r1_AACAACCTCCCATACC, hft_w20_p3_r1_AACAACCTCGTCTACC, hft_w20_p3_r1_AACAAGAAGCCGTTAT, hft_w20_p3_r1_AACAAGACAAACACGG, hft_w20_p3_r1_AACAAGACACATAGCT, hft_w20_p3_r1_AACAAGACACGTCTCT, hft_w20_p3_r1_AACAAGACAGGAGACT, hft_w20_p3_r1_AACAAGAGTCTCAGAT, hft_w20_p3_r1_AACAAGATCGCTAAAC, hft_w20_p3_r1_AACAAGATCGTCCATC, hft_w20_p3_r1_AACACACCACTGCTTC, hft_w20_p3_r1_AACACACCATGACTAC, hft_w20_p3_r1_AACACACGTAGTAAGT, hft_w20_p3_r1_AACACACGTATCTCGA, hft_w20_p3_r1_AACACACTCCTACGAA, hft_w20_p3_r1_AACAGGGAGAGTTGAT, hft_w20_p3_r1_AACAGGGAGATTAGTG, hft_w20_p3_r1_AACAGGGAGCAACAGC, hft_w20_p3_r1_AACAGGGCAGCACAAG, hft_w20_p3_r1_AACAGGGTCGAGAACG, hft_w20_p3_r1_AACAGGGTCTCAATCT, ...]

[57868 rows x 0 columns]

### 2.3. Metadata

In [13]:
meta = pd.read_csv(
    meta_file,
    sep = '\t',
    index_col='Cell.ID'
    )

### Meta data is including the information about patients.
### Put prefix to indicate it
meta = meta.add_prefix('Auth_')

In [14]:
meta

Auth_Sample.ID Auth_Age Auth_Tissue.ID  \
Cell.ID                                                                 
hft_w20_p3_r1_AAACCCAAGCTGCGAA  hft_w20_p3_r1    pcw20           HFT3   
hft_w20_p3_r1_AAACCCAAGGTAGTAT  hft_w20_p3_r1    pcw20           HFT3   
hft_w20_p3_r1_AAACCCACAACTCCAA  hft_w20_p3_r1    pcw20           HFT3   
hft_w20_p3_r1_AAACCCACATAGTCAC  hft_w20_p3_r1    pcw20           HFT3   
hft_w20_p3_r1_AAACCCAGTACAGGTG  hft_w20_p3_r1    pcw20           HFT3   
...                                       ...      ...            ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA  hft_w16_p7_r2    pcw16           HFT7   
hft_w16_p7_r2_TTTGTTGCAGGCTACC  hft_w16_p7_r2    pcw16           HFT7   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA  hft_w16_p7_r2    pcw16           HFT7   
hft_w16_p7_r2_TTTGTTGGTCGTACAT  hft_w16_p7_r2    pcw16           HFT7   
hft_w16_p7_r2_TTTGTTGGTTAGTTCG  hft_w16_p7_r2    pcw16           HFT7   

                               Auth_Sample.Type Auth_Assay Auth_Batch  \
Cell.ID                                                                 
hft_w20_p3_r1_AAACCCAAGCTGCGAA              HFT     RNA v3   b2019_06   
hft_w20_p3_r1_AAACCCAAGGTAGTAT              HFT     RNA v3   b2019_06   
hft_w20_p3_r1_AAACCCACAACTCCAA              HFT     RNA v3   b2019_06   
hft_w20_p3_r1_AAACCCACATAGTCAC              HFT     RNA v3   b2019_06   
hft_w20_p3_r1_AAACCCAGTACAGGTG              HFT     RNA v3   b2019_06   
...                                         ...        ...        ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA              HFT     RNA NG   b2020_03   
hft_w16_p7_r2_TTTGTTGCAGGCTACC              HFT     RNA NG   b2020_03   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA              HFT     RNA NG   b2020_03   
hft_w16_p7_r2_TTTGTTGGTCGTACAT              HFT     RNA NG   b2020_03   
hft_w16_p7_r2_TTTGTTGGTTAGTTCG              HFT     RNA NG   b2020_03   

                               Auth_seurat_clusters  Auth_RNA.Counts  \
Cell.ID                                                                
hft_w20_p3_r1_AAACCCAAGCTGCGAA                  c16             1397   
hft_w20_p3_r1_AAACCCAAGGTAGTAT                  c11            14338   
hft_w20_p3_r1_AAACCCACAACTCCAA                  c17             9260   
hft_w20_p3_r1_AAACCCACATAGTCAC                   c0             4025   
hft_w20_p3_r1_AAACCCAGTACAGGTG                   c4             7131   
...                                             ...              ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA                   c6             8532   
hft_w16_p7_r2_TTTGTTGCAGGCTACC                   c2             6689   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA                   c2             3865   
hft_w16_p7_r2_TTTGTTGGTCGTACAT                   c2             5293   
hft_w16_p7_r2_TTTGTTGGTTAGTTCG                   c6             5291   

                                Auth_RNA.Features  Auth_Percent.MT  ...  \
Cell.ID                                                             ...   
hft_w20_p3_r1_AAACCCAAGCTGCGAA                677         0.070866  ...   
hft_w20_p3_r1_AAACCCAAGGTAGTAT               4301         0.053200  ...   
hft_w20_p3_r1_AAACCCACAACTCCAA               3481         0.043511  ...   
hft_w20_p3_r1_AAACCCACATAGTCAC               1969         0.015396  ...   
hft_w20_p3_r1_AAACCCAGTACAGGTG               2930         0.044690  ...   
...                                           ...              ...  ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA               3497         0.018749  ...   
hft_w16_p7_r2_TTTGTTGCAGGCTACC               2393         0.045577  ...   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA               1799         0.035954  ...   
hft_w16_p7_r2_TTTGTTGGTCGTACAT               2365         0.032861  ...   
hft_w16_p7_r2_TTTGTTGGTTAGTTCG               2530         0.006803  ...   

                                Auth_Cell.Barcode Auth_DF_pANN  \
Cell.ID                                                          
hft_w20_p3_r1_AAACCCAAGCTGCGAA   AAACCCAAGCTGCGAA     0.086339   
hft_w20_p3_r1_AAACCC

### 2.4 Optimizing adata.X data type to uint16 
* optimized data type boosts the process

In [15]:
### Check if adata.X contains non-integer data
### If the data is not big, check all
### If the data is too big, check random data
if adata.shape[0] <= 250000: 
    print('Checking complete adata')
    if np.equal(np.mod(adata.X.toarray(), 1), 0).all() != True:
        print('CAREFUL: non-integer matrix loaded!')
else:
    print ('Checking a subsample of adata')
    if np.equal(np.mod(sc.pp.subsample(adata, n_obs=250000, random_state=0, copy=True).X.toarray(), 1), 0).all() != True:
        print('CAREFUL: non-integer matrix loaded!')

Checking complete adata


In [16]:
### Check if the maximum value in the data fits within the 'uint16' range (0 ~ 65535)
### 32,000 <- safe threshold
maxCount = csr_matrix.max(adata.X)
if maxCount < 32000: 
    print('Change X type to integer')
    adata.X = adata.X.astype('uint16') # Optimization to 'unit16'
    if maxCount != csr_matrix.max(adata.X):
        print('CAREFUL: max count value has changed!')
else: 
    print('X type not changed') # In this case, consider different data type

Change X type to integer


### 2.5 Build adata.obs

In [17]:
if meta.shape[0] != adata.obs.shape[0]:
    print('CAREFUL: expression matrix and metadata size are not coherent!')
    print ('Metadata rows: ' + str(meta.shape[0]))

In [18]:
adata.obs = adata.obs.join(meta, how='left', validate='one_to_one')

### 2.6 Build adata.var

In [19]:
annot = sc.queries.biomart_annotations(
        "hsapiens",
        ["ensembl_gene_id", "external_gene_name","start_position", "end_position", "chromosome_name"],
    ).set_index("ensembl_gene_id")

In [20]:
adata.var = adata.var.join(annot, how='left', validate='one_to_one')
print(adata.var['external_gene_name'].isnull().sum())
del annot

8463


* The value 8463 is different with https://github.com/BrainOmicsCourse/BrainOmics2024/blob/main/1_Day1/Compiled/0_AssembleAdata.html
* I decided to ignore this problem

In [21]:
adata.var['ensg'] = adata.var.index.tolist()
adata.var['external_gene_name'] = adata.var['external_gene_name'].fillna(adata.var.ensg)

In [22]:
#replace external gene name as index
adata.var.index = adata.var['external_gene_name']
adata.var.drop('external_gene_name', axis = 1, inplace = True)
adata.var_names_make_unique()

In [23]:
adata.var.head(3)

start_position  end_position chromosome_name  \
external_gene_name                                                 
MIR1302-2HG                28589.0       31109.0               1   
FAM138A                    34553.0       37595.0               1   
OR4F5                      65419.0       71585.0               1   

                               ensg  
external_gene_name                   
MIR1302-2HG         ENSG00000243485  
FAM138A             ENSG00000237613  
OR4F5               ENSG00000186092

### 2.7. Add additional fields to adata.obs

In [24]:
adata.obs['dataset_id'] = 'Id0003'
adata.obs['sample_id'] = adata.obs['Auth_Sample.ID']
adata.obs['brain_region'] = 'cerebral_cortex'
adata.obs['age'] = adata.obs['Auth_Age'].str.replace('pcw', 'PCW_')
adata.obs['stage'] = 'prenatal'
adata.obs['batch_key'] = adata.obs['Auth_Batch'] + '_' + adata.obs['Auth_Assay'].str.replace(' ', '')

In [25]:
Dict = {'c0': 'ExN_N5', 'c1': 'In_CGE', 'c2': 'ExN_N1',
        'c3': 'In_MGE', 'c4': 'ExN_N4', 'c5': 'ExN_N2', 
        'c6': 'RG_early', 'c7': 'ExN_N7', 'c8': 'CycProg', 
        'c9': 'ExN_N3', 'c10': 'RG_late', 'c11': 'GliaPg', 
        'c12': 'ExN_N6', 'c13': 'SubPlate', 'c14': 'IPC', 
        'c15': 'ExN_N8', 'c16': 'Microglia', 'c17': 'OPC_Oligo', 
        'c18': 'tRG', 'c19': 'Pericytes', 'c20': 'Endo',
        'c21': 'RBC', 'c22': 'VLMC'
       }

adata.obs['cell_label'] = adata.obs['Auth_seurat_clusters'].replace(Dict)

### 2.8. Save

In [26]:
adata

AnnData object with n_obs × n_vars = 57868 × 33355
    obs: 'Auth_Sample.ID', 'Auth_Age', 'Auth_Tissue.ID', 'Auth_Sample.Type', 'Auth_Assay', 'Auth_Batch', 'Auth_seurat_clusters', 'Auth_RNA.Counts', 'Auth_RNA.Features', 'Auth_Percent.MT', 'Auth_Percent.Ribo', 'Auth_Cell.Barcode', 'Auth_DF_pANN', 'Auth_DF_classification', 'Auth_DF_pANN_quantile', 'Auth_Spliced.Counts', 'Auth_Spliced.Features', 'Auth_Unspliced.Counts', 'Auth_Unspliced.Features', 'Auth_Ambiguous.Counts', 'Auth_Ambiguous.Features', 'dataset_id', 'sample_id', 'brain_region', 'age', 'stage', 'batch_key', 'cell_label'
    var: 'start_position', 'end_position', 'chromosome_name', 'ensg'

In [27]:
if isspmatrix(adata.X) == False:
    adata.X = csr_matrix(adata.X)
    print('Converted adata.X to', type(adata.X))

In [28]:
adata.write(output_file, compression='gzip')

In [29]:
os.path.getsize(output_file)

245726691